File: core/src/main/java/heroes/journey/registries/Registry.java
package heroes.journey.registries;

import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class Registry<T extends Registrable> extends HashMap<String,T> {

    public List<T> get(List<String> entryStrings) {
        return entryStrings.stream().map(this::get) // get the Action for each string
            .filter(Objects::nonNull) // optionally skip nulls
            .collect(Collectors.toList());
    }

    @Override
    public T get(Object id) {
        T item = super.get(id);
        if (item == null) {
            System.out.println("Failed to get item with id " + id.toString());
        }
        return item;
    }

    public T register(T entry) {
        if (containsKey(entry.getId())) {
            throw new RuntimeException(this.getClass() + " cannot register with id " + entry.getId() +
                " because that id is already registered");
        }
        put(entry.getId(), entry);
        return entry;
    }
}

File: modlib/src/main/java/heroes/journey/modding/api/registration/Registry.kt
package heroes.journey.modding.api.registration

class Registry<T> : MutableMap<String, T> by mutableMapOf() {
    fun get(entryStrings: List<String>): List<T> {
        return entryStrings.mapNotNull { get(it) }
    }

    fun getOrNull(key: String?): T? {
        if (key == null) return null
        return this[key]
    }

    fun register(entry: T): T {
        val key = entry.toString()
        if (containsKey(key)) {
            throw RuntimeException("${this::class.java} cannot register with name $key because that name is already registered")
        }
        put(key, entry)
        return entry
    }
}

